spatial/r3/mat.go

```diff
@@ -163,3 +166,57 @@ func (m *Mat) VecCol(j int) Vec {
 	}
 	return Vec{X: m.At(0, j), Y: m.At(1, j), Z: m.At(2, j)}
 }
+
+// Outer calculates the outer product of the vectors x and y,
+// where x and y are treated as column vectors, and stores the result in the receiver.
+//  m = alpha * x * yᵀ
+func (m *Mat) Outer(alpha float64, x, y Vec) {
+	ax := alpha * x.X
+	ay := alpha * x.Y
+	az := alpha * x.Z
+	m.Set(0, 0, ax*y.X)
+	m.Set(0, 1, ax*y.Y)
+	m.Set(0, 2, ax*y.Z)
+
+	m.Set(1, 0, ay*y.X)
+	m.Set(1, 1, ay*y.Y)
+	m.Set(1, 2, ay*y.Z)
+
+	m.Set(2, 0, az*y.X)
+	m.Set(2, 1, az*y.Y)
+	m.Set(2, 2, az*y.Z)
+}
+
+// Sets the rotation component of this matrix to the rotation specified by q,
```

Comment:
```
// RotationFromQuat sets the rotation component of this matrix to the rotation specified by q.
//
// See https://en.wikipedia.org/wiki/Rotation_matrix#Quaternion for details.
```

What is the behaviour is q is not a unit quaternion? Should that be documented? Should q be scaled to unit length if it is not?

Reply:
Definitely should be documented, slipped my mind. Should be easy enough to check in the function after calculating x2, y2, z2 and panic if not normalized or normalize then and there. I think normalizing non-unit quaternions is OK though I'm unsure of whether it's good design.

Reply:
You can just call `quat.Abs(q)` to find if it's unit. The question is whether people want to scale at the same time.

Reply:
... and `quat.Scale(1/quat.Abs(q). q)`.

Reply:
What about having a boolean argument `scale` to make it more explicit? Think it might be worthwhile since it works as explicit documentation.

Reply:
TBH, I think just documenting that no quaternion scaling takes place is fine, this favours composition over parameterisation.